### PR TITLE
[fix bug]1. crash when create mesh with _struct undefined, 2. when get property from material

### PR DESCRIPTION
**Pull request by timlyeee**
to fix: https://github.com/cocos/3d-tasks/issues/12716 https://github.com/cocos/3d-tasks/issues/12732 

* check if _struct is undefined or empty, both valid.
* fix get property error when no passIdx is given, use default method.
* fix std::variant to private object
* fix get property failed when try to get from Vec2 or Color etc.

no public api changed

### CODE REVIEW

**Review by dumganhar — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Comment by timlyeee** [comment]
Add bug fix for issue https://github.com/cocos/3d-tasks/issues/12732 when entering to scene Physics

File: cocos/core/assets/material.jsb.ts [hunk 1]
```diff
@@ -156,31 +166,36 @@ matProto.setProperty = function (name: string, val: MaterialPropertyFull | Mater
 };
 
 matProto.getProperty = function (name: string, passIdx?: number) {
-    const val = this._getProperty(name, passIdx);
+    let val: any;
+    if (passIdx) {
```
Comment (PatriceJiang) [inline cocos/core/assets/material.jsb.ts]: `passIdx !== undefined` ?

Comment (timlyeee) [reply cocos/core/assets/material.jsb.ts]: undefine or null both make it no sense, so I suppose that `if(passIdx)` is ok here

Comment (PatriceJiang) [reply cocos/core/assets/material.jsb.ts]: Should zero be treated as undefined?

File: cocos/core/assets/material.jsb.ts [hunk 2]
```diff
@@ -88,7 +88,17 @@ function wrapSetProperty (cb: setProperyCB, target: Material, name: string, val:
         cb.call(target, name, val);
     }
 }
-
+enum  MathType {
```
Comment (dumganhar) [inline cocos/core/assets/material.jsb.ts]: Remove a space between enum and MathType.

**Review by minggo — APPROVED** [review]

File: native/cocos/bindings/manual/jsb_conversions_spec.cpp [hunk 3]
```diff
@@ -133,6 +134,17 @@ enum class DataType {
     INT,
     FLOAT
 };
+enum class MathType {
```
Comment (dumganhar) [inline native/cocos/bindings/manual/jsb_conversions_spec.cpp:136]: Add a `NOTE` comment that MathType here should synchronize which in material.jsb.ts.

Comment (timlyeee) [reply native/cocos/bindings/manual/jsb_conversions_spec.cpp:136]: I guess the note should be added in material.jsb.ts? that's more logical

File: cocos/core/assets/material.jsb.ts (hunk 2 above)
Comment (timlyeee) [reply cocos/core/assets/material.jsb.ts]: 1

File: cocos/core/assets/material.jsb.ts (hunk 1 above)
Comment (timlyeee) [reply cocos/core/assets/material.jsb.ts]: passIdx might be 0, right about that

File: cocos/3d/assets/mesh.jsb.ts [hunk 4]
```diff
@@ -63,7 +63,10 @@ Object.defineProperty(meshAssetProto, 'struct', {
 });
 
 meshAssetProto.onLoaded = function () {
-    this.setStruct(this._struct);
+    // might be undefine
```
Comment (dumganhar) [inline cocos/3d/assets/mesh.jsb.ts]: undefine -> undefined

Comment (timlyeee) [reply cocos/3d/assets/mesh.jsb.ts]: 1

File: cocos/core/assets/material.jsb.ts [hunk 5]
```diff
@@ -156,31 +167,36 @@ matProto.setProperty = function (name: string, val: MaterialPropertyFull | Mater
 };
 
 matProto.getProperty = function (name: string, passIdx?: number) {
-    const val = this._getProperty(name, passIdx);
+    let val: any;
+    if (passIdx != undefined) {
```
Comment (PatriceJiang) [inline cocos/core/assets/material.jsb.ts]: Should use `!==` here

Comment (timlyeee) [reply cocos/core/assets/material.jsb.ts]: 1

**Review by PatriceJiang — APPROVED** [review]

**Pull request merged**